### DISCUSSION TITLE
Hitting `ThrottlingException` on `GetWorkGroup` with threads turned up

### DISCUSSION THREAD

**Discussion by dacreify** [dbt-athena/dbt-athena#591] [General]
Recently we started hitting `ThrottlingException` on `GetWorkGroup` calls at this spot in the code:

https://github.com/dbt-athena/dbt-athena/blob/a1b8db5de90b20557bcd5e0c51a30177bcddaa5f/dbt/adapters/athena/impl.py#L231

From CloudTrail it looks like `dbt-athena` winds up making a `GetWorkGroup` call for every model run. There's no documented quota for this call but obviously we're hitting one.  Actual `StartQueryExecution` can go 20/second or burst to 80/second so `GetWorkGroup` definitely seems to be below that in any case.

Anyone else hit this? Can we cache the results of `GetWorkGroup` to avoid it?

**Comment by nicor88** [comment]
Thanks for your issue, caching could make sense. We can consider to put a caching wrapper on an inner function that contains
```
if creds.work_group:
            work_group = athena_client.get_work_group(WorkGroup=creds.work_group)
            output_location = (
                work_group.get("WorkGroup", {})
                .get("Configuration", {})
                .get("ResultConfiguration", {})
                .get("OutputLocation", None)
            )

            output_location_enforced = (
                work_group.get("WorkGroup", {}).get("Configuration", {}).get("EnforceWorkGroupConfiguration", False)
            )

            return output_location is not None and output_location_enforced
        else:
            return False
```

How many models do you have? I'm wondering which scale you have that lead you to hit this throttling.
We can even consider to add a retry wrapper with backoff on top of caching.

**Reply by dacreify** [reply]
The selector we're hitting this with has around 90 models and we have `threads` at 50. There's no requirement for this level of concurrency yet (was more of a stress test) so we can just turn it down for now. In the future our workload may need that capacity, though, so it'd be good for `GetWorkGroup` not to be the bottleneck.

AWS support acknowledge there is no documented quota for these calls but is checking further internally.

**Reply by dacreify** [reply]
Also looks like retries are already happening -- maybe within `boto3`? Here's the full error message:

`An error occurred (ThrottlingException) when calling the GetWorkGroup operation (reached max retries: 4): Rate exceeded`

**Reply by nicor88** [reply]
using 50 as thread count is really high, that's what lead to the issue. You hit GetWorkGroup api operation because you have at max 50 times at the same time. boto3 under the hood retry already indeed, but we can cadd a re-try wrapper that can be configured via num_boto3_retries eventually. Anyhow, even if GetWorkGroup succeeded at that scale, 50 used as thread you will start to see athena resource exhausted errors, or queries in queue mode...you solve an issue, but you will face definitely another issue.

**Reply by dacreify** [reply]
I agree that the high concurrency is what provoked the issue, but also it doesn't seem like `GetWorkGroup` should be the bottleneck on how far you can scale a `dbt-athena` setup. The default quota for concurrently running Athena queries is 200, which you can launch at a rate of 20/second bursting to 80/s. I can see 50 threads being a valid configuration for a large project given these limits. i.e. The run kicks off by launching 50 queries at once, then settles in to a lower rate as models take different amounts of time to complete and new ones are begun.

Anecdotally we were running with 50 threads for months with no model failures. (Given the charges are per query, we thought, why not run it as fast as it will go?) Maybe retries were just barely masking the `GetWorkGroup` throttling. 🤔 

Would you agree that we should still file an issue to add caching for `GetWorkGroup`?

**Reply by juliansteger-sc** [reply]
we started receiving the same `ThrottlingException` quite recently, so pretty sure AWS changed some internal quota:

```
An error occurred (ThrottlingException) when calling the GetWorkGroup operation (reached max retries: 2): Rate exceeded
```
We didn't change anything with our thread count recently.

So, I think we should optimize the number of api calls / cache the workgroup information

**Reply by nicor88** [reply]
@dacreify I forgot to come back to you, yes let's add a caching as you suggested. ~Do you want to take care of it or shall I?~ - if this is not enough, we will add a retry wrapper as well.

**Reply by nicor88** [reply]
@dacreify let's try this https://github.com/dbt-athena/dbt-athena/pull/596
if this caching is not enough, the only way that I can think of is to add a retry wrapper with backoff...
EDIT: the PR was merged, and it's in the main branch. 
@dacreify and @juliansteger-sc could you try to install the adapter from a specific git commit and see if the issue is resolved?

**Reply by dacreify** [reply]
Thanks, @nicor88!

AWS acknowledged that this was a regression on our support request with them:

> We saw an increase in throttling exceptions in the Athena service. Throttling exceptions occur when a user exceeds the limit for the number of Athena API calls allowed per second. Upon investigation, we discovered an error in how the API call rate was calculated. To address this, we have deployed a fix across all our regions on 03/12/2024. If you are still experiencing throttling exceptions or any related issues, we kindly ask you to reach out to us.

They told me what the rate limit on `GetWorkGroup` is but I think they're asking me not to disclose it. I will hazard to share that it _is_ lower than the default for `StartQueryExecution` so this enhancement will still be a help for scaling.

**Reply by nicor88** [reply]
@dacreify Did you verify that the current caching implementation allow you to reduce the Throttling exceptions?

**Reply by dacreify** [reply]
Left results on the issue: https://github.com/dbt-athena/dbt-athena/issues/595#issuecomment-2000608564
